TCS Pivotted to using industry standard CVs for extended decoder iden…

TCSDCC-Jordan opened the pull request:
…tification

After continuing to review products at a rapid rate TCS has chosen not to use a software hash, but rather a manufacturer controlled value in CV's 110 and 111 to handle advanced identification for our expanding sound products.

We will likely tweak software frequently, but rarely change the sound set for a given decoder. By locking this number to a manually controlled value we can insure minor bug changes dont require new JMRI definitions reducing the overall size of TCS product definitions in JMRI.

danielb987 commented:
@dheap 
Can you take a look at this PR as well? Thanks.

TCSDCC-Jordan commented:
I'm so sorry. I've spotted a bug in this in more rigorous testing today.
Can we cancel this PR so I can address this issue?

My apologies for headache

Best Regards,
Jordan McBride
Train Control Systems
www.tcsdcc.com

On Wed, Jan 12, 2022 at 9:37 AM Daniel ***@***.***> wrote:

> @dheap <https://github.com/dheap>
> Can you take a look at this PR as well? Thanks.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/JMRI/JMRI/pull/10568#issuecomment-1011109851>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AF3PQEQ6SF4NDWWD5IJJM6DUVWGZ5ANCNFSM5LXMAOPQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

geowar1 commented:
Just edit the name with the "WIP:" prefix so everyone knows it's a Work In Progress.

Sent from GeoWar’s iPad Pro

> On Jan 12, 2022, at 8:16 AM, TCSDCC-Jordan ***@***.***> wrote:
> 
> ﻿
> I'm so sorry. I've spotted a bug in this in more rigorous testing today.
> Can we cancel this PR so I can address this issue?
> 
> My apologies for headache
> 
> Best Regards,
> Jordan McBride
> Train Control Systems
> www.tcsdcc.com
> 
> On Wed, Jan 12, 2022 at 9:37 AM Daniel ***@***.***> wrote:
> 
> > @dheap <https://github.com/dheap>
> > Can you take a look at this PR as well? Thanks.
> >
> > —
> > Reply to this email directly, view it on GitHub
> > <https://github.com/JMRI/JMRI/pull/10568#issuecomment-1011109851>, or
> > unsubscribe
> > <https://github.com/notifications/unsubscribe-auth/AF3PQEQ6SF4NDWWD5IJJM6DUVWGZ5ANCNFSM5LXMAOPQ>
> > .
> > Triage notifications on the go with GitHub Mobile for iOS
> > <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> > or Android
> > <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
> >
> > You are receiving this because you authored the thread.Message ID:
> > ***@***.***>
> >
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android. 
> You are receiving this because you are subscribed to this thread.

danielb987 commented:
No need to be sorry for this! It has happen to me multiple times.

Either you can close it, or you can add WIP to the title to mark it as 
Work In Progress, as George Warner suggested.

Thanks for your work on this!
Daniel

2022-01-12 17:16 skrev TCSDCC-Jordan:
> I'm so sorry. I've spotted a bug in this in more rigorous testing
> today.
> Can we cancel this PR so I can address this issue?
> 
> My apologies for headache
> 
> Best Regards,
> Jordan McBride
> Train Control Systems
> www.tcsdcc.com
> 
> On Wed, Jan 12, 2022 at 9:37 AM Daniel ***@***.***> wrote:
> 
>> @dheap <https://github.com/dheap>
>> Can you take a look at this PR as well? Thanks.
>> 
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/JMRI/JMRI/pull/10568#issuecomment-1011109851>,
> or
>> unsubscribe
>> 
> <https://github.com/notifications/unsubscribe-auth/AF3PQEQ6SF4NDWWD5IJJM6DUVWGZ5ANCNFSM5LXMAOPQ>
>> .
>> Triage notifications on the go with GitHub Mobile for iOS
>> 
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
>> or Android
>> 
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>> 
>> You are receiving this because you authored the thread.Message ID:
>> ***@***.***>
>> 
> --
> Reply to this email directly, view it on GitHub [1], or unsubscribe
> [2].
> You are receiving this because you commented.Message ID:
> ***@***.***>
> 
> Links:
> ------
> [1] https://github.com/JMRI/JMRI/pull/10568#issuecomment-1011216075
> [2] 
> https://github.com/notifications/unsubscribe-auth/AE2REVLTUDM54JH7NBJHZZTUVWSPLANCNFSM5LXMAOPQ

TCSDCC-Jordan commented:
@dheap @danielb987 

This is no longer WIP and is ready for review for pull-in. Thanks again for your understanding and input to get me on the right track.

TCSDCC-Jordan commented:
Ignore previous (now deleted) comment. Changes coming from higher up than me.

danielb987 commented:
> Ignore previous (now deleted) comment. Changes coming from higher up than me.

No problem. Thanks for your work on this.

dheap commented:
@TCSDCC-Jordan I haven't had time to review this code in detail and presently am juggling around severe electrical storms and extended family staying with me.

At first glance, what concerns me is there appears to be a net reduction of code in these two files. We have to continue to support the existing ID procedure for older decoders still in the field. Likewise I'd expect to see at least one extra Test, not simply modification of an existing test.

You are in the best position to check with real hardware whether the new code not only correctly identifies a new V5 decoder but also continues to identify some existing decoder models.

TCSDCC-Jordan commented:
Alright, I have pushed up a final (in terms of TCS needs) version of this.

@dheap Thanks for taking time to read through and make a note of this change. This is a result of internal TCS chaos which I hope I can shed some light on. Briefly, over the summer, there was an idea that we would use github code hashes for truly unique productID's in decoder pro. The problem with that, which I am sure you see already, is those change all of the time. During this window of time we did not achieve our goal of shipping V5 and so ultimately, no products went out where the primary mode of identification was the github hash.  Additionally, no decoder definition exists in JMRI for TCS at this point with a github hash as it's productID so we are very safe on losing these CV lookups.

You may note the last change to the TCS portion of this file was my PR some months back.

In this updated revision, and actually the previous as well, but more importantly on this algorithm we account for:
1. Non-sound/very old decoders (they fall through into standard lookups at that test for value < 129 (value being the CV 249 value).
2. Sound, but pre 4.5 or 5 versions ( this is that fairly complex if statement in the last test) also fall through back to the previous identify style because CV 7 will never be greater than 5 on anything pre V5 except an odd case in steam which was just released which is the first half of that ugly if statement.

I have added an additional test specifically focused on that complex if statement below. So we now have tests for:

- Sound V5 and up (the one new test you had seen)
- Sound V4 and down (the text I just added)
- Mobile decoders (this was mike moshers "testIdentifyTCSV4()" which I have renamed to "testIdentifyTCSMobile()" because a value of 80 in CV 249 is explicitly a non-sound decoder and V4 does not apply to non-sound (we have been over 40 in CV 7 on mobile for close to 20 years now).

Do you see another test that could be of value here? Thanks in advance!

dheap commented:
I see your problem.
There's no problem with having multiple ProductIDs for one model (comma separated list), as we do with the ESU range, but the ProductID only changes for board hardware revisions, not firmware code version or sound project number.

As long as you're satisfied that the IdentifyDecoder routine covers all fIeld-released TCS model variants, I'm happy to approve this.

ESU uses the S9.3.2 section 4.2.4 RailCom CVs ProductID field (as well as other fields) for decoder information storage.

TCSDCC-Jordan commented:
I, and the rest of the engineering team at TCS who looked over the new algo, are comfortable with this coverage so we are good to go.

Thanks for taking the time to give this real consideration along the way. That is really helpful to pause and make sure we have considered all variants.

dheap reviewed (approved):
I see no serious problems and @TCSDCC-Jordan has obviously done his homework thoroughly.
